From ac61625c7fa15a3157a107bdd80036c7dc58f709 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Fri, 21 Jul 2023 15:15:00 +0800
Subject: [PATCH 13/53] [Huawei] mm/sharepool: Process the permission in
 sp_mmap()

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Use spa->flags to determine the permission in sp_mmap() and this would
cover all the callpath.

The new task of a group should not have write access to a exist spa that
has allocate with SP_PROT_RO. And SP_PROT_FOCUS matter nothing but the
virtual address of the spa located at a special range. We just delete
it.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 26 +++++++-------------------
 1 file changed, 7 insertions(+), 19 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index f26aa71b58f2..2c1951e79a7a 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -1463,10 +1463,6 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 		unsigned long populate = 0;
 		struct file *file = spa_file(spa);
 		unsigned long addr;
-		unsigned long prot_spa = prot;
-
-		if ((spa->flags & (SP_PROT_RO | SP_PROT_FOCUS)) == (SP_PROT_RO | SP_PROT_FOCUS))
-			prot_spa &= ~PROT_WRITE;
 
 		__sp_area_drop_locked(prev);
 		prev = spa;
@@ -1479,7 +1475,7 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 		spin_unlock(&sp_area_lock);
 
 		if (spa->type == SPA_TYPE_K2SPG && spa->kva) {
-			addr = sp_remap_kva_to_vma(spa, mm, prot_spa, NULL);
+			addr = sp_remap_kva_to_vma(spa, mm, prot, NULL);
 			if (IS_ERR_VALUE(addr))
 				pr_warn("add group remap k2u failed %ld\n", addr);
 
@@ -1495,7 +1491,7 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 			continue;
 		}
 
-		addr = sp_mmap(mm, file, spa, &populate, prot_spa);
+		addr = sp_mmap(mm, file, spa, &populate, prot);
 		if (IS_ERR_VALUE(addr)) {
 			up_write(&mm->mmap_lock);
 			pr_warn("add group: sp mmap failed %ld\n", (long)addr);
@@ -2086,6 +2082,9 @@ static unsigned long sp_mmap(struct mm_struct *mm, struct file *file,
 	unsigned long pgoff = addr_offset(spa) >> PAGE_SHIFT;
 	struct vm_area_struct *vma;
 
+	if (spa->flags & SP_PROT_RO)
+		prot &= ~PROT_WRITE;
+
 	atomic_inc(&spa->use_count);
 	addr = __do_mmap_mm(mm, file, addr, size, prot, flags, vm_flags, pgoff,
 			 populate, NULL);
@@ -2212,13 +2211,11 @@ static long sp_mbind(struct mm_struct *mm, unsigned long start, unsigned long le
 
 #define SP_SKIP_ERR 1
 static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
-	struct sp_group_node *spg_node, struct sp_alloc_context *ac,
+	unsigned long prot, struct sp_alloc_context *ac,
 	nodemask_t *nodemask)
 {
 	int ret;
 	unsigned long mmap_addr;
-	/* pass through default permission */
-	unsigned long prot = PROT_READ | PROT_WRITE;
 	unsigned long populate = 0;
 
 	down_write(&mm->mmap_lock);
@@ -2228,12 +2225,6 @@ static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
 		return SP_SKIP_ERR;
 	}
 
-	if (spg_node)
-		prot = spg_node->prot;
-
-	if (ac->sp_flags & SP_PROT_RO)
-		prot = PROT_READ;
-
 	/* when success, mmap_addr == spa->va_start */
 	mmap_addr = sp_mmap(mm, spa_file(spa), spa, &populate, prot);
 	if (IS_ERR_VALUE(mmap_addr)) {
@@ -2297,7 +2288,7 @@ static int sp_alloc_mmap_populate(struct sp_area *spa,
 	/* create mapping for each process in the group */
 	list_for_each_entry(spg_node, &spa->spg->proc_head, proc_node) {
 		mm = spg_node->master->mm;
-		mmap_ret = __sp_alloc_mmap_populate(mm, spa, spg_node, ac, &__nodemask);
+		mmap_ret = __sp_alloc_mmap_populate(mm, spa, spg_node->prot, ac, &__nodemask);
 		if (mmap_ret) {
 			/*
 			 * Goto fallback procedure upon ERR_VALUE,
@@ -2472,9 +2463,6 @@ static unsigned long sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *
 		goto put_mm;
 	}
 
-	if (kc && (kc->sp_flags & SP_PROT_RO))
-		prot = PROT_READ;
-
 	ret_addr = sp_mmap(mm, spa_file(spa), spa, &populate, prot);
 	if (IS_ERR_VALUE(ret_addr)) {
 		pr_debug("k2u mmap failed %lx\n", ret_addr);
-- 
2.17.1

